File: PROGRAMMING/MATLAB CODE/magnesiumsensor.m
clc;
clear;
close all;
%1.Simulate Lead sensor data
fs=10;
t=0:1/fs:100;
true_manganese=10+2*sin(0.1*t)+0.4*randn(size(t));
%Add random sensor noise(simulating measurement error)
noise=0.8*randn(size(t));
manganese_noisy=true_manganese+noise;
%2.Moving Average Filter
window_size=5; %Number of samples in windows
b=(1/window_size)*ones(1,window_size);
a=1;
manganese_moving_avg=filter(b,a,manganese_noisy);
%3.Median Filter
windowSize_med=5;
manganese_median=medfilt1(manganese_noisy,windowSize_med);
%4.Butterworth Low-pass Filter
cutoff_freq=0.1; %Normalized cutoff frequency
[b_butter,a_butter]=butter(4,cutoff_freq);
manganese_butter=filter(b_butter,a_butter,manganese_noisy);
%5.Plot results
figure('Name','Manganese Sensor Data Filtering','NumberTitle','off');
subplot(4,1,1);
plot(t,manganese_noisy,'r');
title('Unfiltered Manganese Sensor Data(Noisy)');
xlabel('Time (s)');
ylabel('Manganese Concentration (ppb)');
grid on;
subplot(4,1,2);
plot(t,manganese_moving_avg,'b');
title('Manganese Sensor Data - Moving Average Filter');
xlabel('Time (s)');
ylabel('Manganese Concentration (ppb)');
grid on;
subplot(4,1,3);
plot(t,manganese_median,'g');
title('Filtered with Median filter');
xlabel('Time (s)');
ylabel('Manganese Concentration (ppb)');
grid on;
subplot(4,1,4);
plot(t,manganese_butter,'m');
title('Filtered with Butterworth Low-Pass Filter (filtfilt)');
xlabel('Time (s)');
ylabel('Manganese Concentration (ppb)');
grid on;
sgtitle('Water Treatment:Manganese Sensor Filtering Comparison');
xlabel('Time (s)');
ylabel('Manganese Concentration (ppb)');
grid on;

File: PROGRAMMING/PYTHON CODE/magnesiumsensor.py
import numpy as np
import matplotlib.pyplot as plt
from scipy.signal import butter, filtfilt, medfilt
# 1. Simulate Magnesium (Mg²⁺) sensor data (ppb concentration)
np.random.seed(42)
time = np.linspace(0, 60, 300)  # 60 seconds,
samples=300
true_magnesium_ppb = 10 + 2 * np.sin(0.3 *  np.pi * time / 10)  # base clean signal
noise = np.random.normal(0, 0.5, size=time.shape)  # random noise
unfiltered_data = true_magnesium_ppb + noise  # simulated raw sensor readings
# 2. Moving Average Filter  
def moving_average(data, window_size=5):
    return np.convolve(data, np.ones(window_size) / window_size, mode='same')
moving_avg_data = moving_average(unfiltered_data, window_size=5)
# 3. Median Filter
median_filtered_data = medfilt(unfiltered_data, kernel_size=5)
# 4. Butterworth Low-Pass Filter
def butter_lowpass_filter(data, cutoff, fs, order=5):
    nyq = 0.5 * fs
    normal_cutoff = cutoff / nyq
    b, a = butter(order, normal_cutoff, btype='low', analog=False)
    return filtfilt(b, a, data)
fs = 300 / 60  # sampling frequency = 300 samples / 60 sec = 5 Hz
cutoff = 0.5   # cutoff frequency (Hz)
butter_filtered_data = butter_lowpass_filter(unfiltered_data, cutoff, fs, order=4)
# 5. Visualization
plt.figure(figsize=(12, 8))
plt.plot(time, unfiltered_data, label='Unfiltered (Raw)', color='gray', alpha=0.6)
plt.plot(time, moving_avg_data, label='Moving Average Filtered', linewidth=2)
plt.plot(time, median_filtered_data, label='Median Filtered', linewidth=2)
plt.plot(time, butter_filtered_data, label='Butterworth Filtered', linewidth=2)
plt.title("Magnesium (Mg²⁺) Sensor Data — Water Treatment Filtration Comparison", fontsize=14)
plt.xlabel("Time (seconds)", fontsize=12)
plt.ylabel("Magnesium Concentration (ppb)", fontsize=12)
plt.legend()
plt.grid(True)
plt.tight_layout()
plt.show()
# 6. Comparison Summary
print("=== Magnesium Sensor Data Filtration Comparison Summary ===")
print(f"Unfiltered Mean Mg²⁺: {np.mean(unfiltered_data):.4f} ppb")
print(f"Moving Average Filtered Mean Mg²⁺: {np.mean(moving_avg_data):.4f} ppb")
print(f"Median Filtered Mean Mg²⁺: {np.mean(median_filtered_data):.4f} ppb")
print(f"Butterworth Filtered Mean Mg²⁺: {np.mean(butter_filtered_data):.4f} ppb")